### PR TITLE
feat(platform): add support for via nodes

### PR DESCRIPTION
**Pull request by shirshanka**
## What are 'Via Nodes'
`Via nodes` are entities designed to represent relationships between two distinct entities in the graph model. They typically represent processes like queries or data jobs that generate column-level lineage between fields, enhancing the model's capabilities in lineage representation. Via nodes are convenient performance and metadata representation shortcuts in cases where paths between two entities transit through a common node that is an intersection point for other paths. 
e.g.  consider fine-grained lineage produced by a job J that looks like

`T1 col c11, c12 -->  (job J) --> T2 col c21, c22`

and

`T1 col c13, c14 --> (job J) --> T2 col 23`

via nodes allow us to concisely represent and query these paths, without needing to create edge specific nodes (e.g. in this case, we do NOT need to mint fake ColumnProcess nodes J-1 , and J-2 to keep the paths separated unlike other metadata systems like Apache Atlas)

Caveat emptor: `via` nodes are experimental - and we'll revisit this decision in case this optimization doesn't yield as many benefits as we think it will.

### Summary
This pull request implements 'via nodes' within the graph model and updates the annotation language for compatibility, aiming to improve the model's data representation and querying capabilities.

### Model Changes
- **Query Support**: Added support for query nodes in both fine-grained and regular lineage models.

### Implementation Changes in GMS
- **Graph Storage and Querying**: Implemented storing and querying of 'via nodes' in the graph, specific to the Elastic search implementation.
- **Enhanced Search-across-Lineage**: Updated the search-across-lineage feature to support multi-path walks, extending beyond shortest path walks.
- **GraphQL Changes**: Made adjustments in GraphQL to accommodate changes in lineage results and search flags due to the new 'via nodes' functionality.
- **Bug Fix - Dataset and DataJob Deletions**: Addressed a bug related to deletions of datasets or dataJobs where relationships involving schemaFields were not properly deleted. This is resolved through the introduction of 'edgeLifecycleOwner' in the graph store.

### Testing
- **Lineage Query Pathway Tests**: Added unit and integration (smoke) tests for the new lineage query pathways to ensure functionality

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-models/src/main/pegasus/com/linkedin/metadata/query/GroupingCriterion.pdl [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+*
+**/
+
+record GroupingCriterion {
+
+    /**
+    * The type of the entity to be grouped.
+    * e.g. schemaField
+    */
+    rawEntityType: string
```
Comment (jjoyce0510) [inline metadata-models/src/main/pegasus/com/linkedin/metadata/query/GroupingCriterion.pdl:13]: one optimization on model: 

make this optional, which means "applies to all entity types"

File: datahub-graphql-core/src/main/resources/search.graphql [hunk 2]
```diff
@@ -1303,4 +1315,30 @@ input SortCriterion {
   The order in which we will be sorting
   """
   sortOrder: SortOrder!
+}
+
+"""
+A grouping specification for search results
+"""
+input GroupingSpec {
+
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/resources/search.graphql:1324]: nit: Can we copy the PDL comment here

File: datahub-graphql-core/src/main/resources/search.graphql [hunk 3]
```diff
@@ -143,6 +143,11 @@ input  SearchFlags {
   Whether to request for search suggestions on the _entityName virtualized field
   """
   getSuggestions: Boolean
+
+  """
+  Additional grouping specifications to apply to the search results
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/resources/search.graphql:148]: Lets add maybe note: 
Notice: This API is experimental and subject to change.

File: datahub-graphql-core/src/main/resources/search.graphql [hunk 4]
```diff
@@ -1303,4 +1315,30 @@ input SortCriterion {
...
+
+"""
+A single grouping criterion for grouping search results
+"""
+input GroupingCriterion {
+
+  """
+  The raw entity type that needs to be grouped
+  """
+  rawEntityType: String!
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/resources/search.graphql:1337]: EntityType -> GraphQL entity type enum 

here is an example:

https://github.com/datahub-project/datahub/blob/master/datahub-graphql-core/src/main/resources/search.graphql#L155

Conversion: 

https://github.com/datahub-project/datahub/blob/master/datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/search/SearchAcrossLineageResolver.java#L58

also, maybe we can call this baseEntityType

File: datahub-graphql-core/src/test/java/com/linkedin/datahub/graphql/resolvers/search/SearchResolverTest.java [hunk 5]
```diff
@@ -45,7 +45,8 @@ public void testDefaultSearchFlags() throws Exception {
             .setSkipAggregates(false)
             .setSkipHighlighting(true) // empty/wildcard
             .setMaxAggValues(20)
-            .setSkipCache(false));
+            .setSkipCache(false)
+            .setConvertSchemaFieldsToDatasets(true));
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/test/java/com/linkedin/datahub/graphql/resolvers/search/SearchResolverTest.java:49]: Is this flag intentional

File: metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/GraphRelationshipMappingsBuilder.java [hunk 6]
```diff
@@ -7,16 +7,23 @@
 
 @Slf4j
 public class GraphRelationshipMappingsBuilder {
+  public static final String EDGE_FIELD_SOURCE = "source";
+  public static final String EDGE_FIELD_DESTINATION = "destination";
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/GraphRelationshipMappingsBuilder.java:11]: ty!

File: metadata-io/src/main/java/com/linkedin/metadata/service/UpdateIndicesService.java [hunk 7]
```diff
@@ -357,12 +374,23 @@ private Pair<List<Edge>, HashMap<Urn, Set<String>>> getEdgesAndRelationshipTypes
...
+      updateFineGrainedEdgesAndRelationships(
+          urn,
+          upstreamLineage.getFineGrainedLineages(),
+          edgesToAdd,
+          urnToRelationshipTypesBeingAdded);
+    } else if (aspectSpec.getName().equals(Constants.INPUT_FIELDS_ASPECT_NAME)) {
       final InputFields inputFields = new InputFields(aspect.data());
       updateInputFieldEdgesAndRelationships(
           urn, inputFields, edgesToAdd, urnToRelationshipTypesBeingAdded);
+    } else if (aspectSpec.getName().equals(Constants.DATA_JOB_INPUT_OUTPUT_ASPECT_NAME)) {
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/service/UpdateIndicesService.java:387]: special cases... we'll need to come back on all of these

**Review by jjoyce0510 — COMMENTED** [review]
Minor naming comments but otherwise that looks good

File: metadata-io/src/main/java/com/linkedin/metadata/search/LineageSearchService.java [hunk 8]
```diff
@@ -206,14 +218,18 @@ public LineageSearchResult searchAcrossLineage(
       }
     }
 
-    // set schemaField relationship entity to be its reference urn
-    LineageRelationshipArray updatedRelationships = convertSchemaFieldRelationships(lineageResult);
-    lineageResult.setRelationships(updatedRelationships);
+    if (SearchUtils.convertSchemaFieldToDataset(searchFlags)) {
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/search/LineageSearchService.java:221]: Is this flag still relevant?

**Review by jjoyce0510 — APPROVED** [review]
Once Ci is green, I think we are good to ship.

File: metadata-io/src/main/java/com/linkedin/metadata/search/LineageSearchService.java (hunk 8 above)
Comment (shirshanka) [reply metadata-io/src/main/java/com/linkedin/metadata/search/LineageSearchService.java:221]: yeah this is just a helper method to keep the check constrained

**Pull request merged**